src/instructlab/model/full_train.py

```diff
@@ -0,0 +1,319 @@
+    # adafactor and gradient checkpointing are memory friendly, we opt to use these in the CPU/MPS loop to fit 7b models.
+    optimizer = Adafactor(
+        model.parameters(), lr=1e-5, scale_parameter=True, relative_step=False
+    )
+    model.gradient_checkpointing_enable()
+
+    model.train()
+    global_step = 1
+
+    # For each epoch do the following:
+    # 1. Incremement Loading Bar
+    # 2. put zeros onto the proper device so we can store some values we need
+    # 3. put each part of the batch onto the device as well using fp6
+    # 4. Forward pass on the model. If using half, we need to do some special handling. If not, then just put the batch onto the device using fp32
+    # 5. Calculate the loss
+    # 6. backward pass on the loss
+    # 7. We are using gradient accumulation to be memory efficient, so if this step is divisible by 4, take a step and zero out the gradients (freeing memory)
+    # 8. clear out some caches if on MPS
+    # 9. Incrememnt global step and loading bar
+    for epoch in range(train_args.num_epochs):
+        dataloader.batch_sampler.set_epoch(epoch)
+        inner_pb = tqdm(range(len(dataloader)), desc=f"Epoch {epoch}")
+        aggregated_values = torch.zeros(3, dtype=torch.float16).to(dev)
+
+        for step, batch in enumerate(dataloader):
+            aggregated_values[0] = batch.pop("num_loss_counted_tokens")
+            aggregated_values[1] = len(batch["input_ids"])
+
+            # Move and cast batch data to device
+            for k in batch:
+                if total_ram < 16:
+                    if k in ["input_ids", "attention_mask"]:
+                        # these two need to be of type long if using .half()
+                        batch[k] = batch[k].to(device=dev, dtype=torch.long)
+                    else:
+                        batch[k] = batch[k].to(device=dev, dtype=torch.float16)
+                else:
+                    batch[k] = batch[k].to(device=dev)
+
+            output = model(**batch, use_cache=False)
+            loss = output.loss
+            aggregated_values[2] = loss.item()
+
+            num_loss_counted_tokens = aggregated_values[0]
+            loss = loss / num_loss_counted_tokens
+
+            loss = loss / accum  # Scale the loss for accumulation steps
+
+            logger.info(
+                f"\nEpoch: {epoch}, Step: {global_step}, Rank: 0, loss = {loss}"
+            )
+
+            # Gradient accumulation
+            loss.backward()  # Backward pass
+
+            # Clear cache before optimizer step
+            if dev.type == "mps":
+                torch.mps.empty_cache()
+
+            # if we are on a step which is divisible by 4, step and zero gradients
+            if (step + 1) % accum == 0:
+                optimizer.step()  # Optimizer step
+                optimizer.zero_grad()  # Zero gradients
+
+                # Clear cache after optimizer step
+                if dev.type == "mps":
+                    torch.mps.empty_cache()
```

Comment:
would be helpful to add some context for why this is necessary, and why only for `mps` and not for `cpu`